feather/common/src/world_source/generating.rs

```diff
@@ -0,0 +1,67 @@
+use std::sync::Arc;
+
+use base::ChunkPosition;
+use flume::{unbounded, Receiver, Sender};
+use worldgen::{ComposableGenerator, WorldGenerator};
+
+use super::{ChunkLoadResult, LoadedChunk, WorldSource};
+
+pub struct GeneratingWorldSource {
+    send: Sender<ChunkPosition>,
+    recv: Receiver<LoadedChunk>,
+}
+
+impl Default for GeneratingWorldSource {
+    fn default() -> Self {
+        Self::new(ComposableGenerator::default_with_seed(42))
+    }
+}
+
+impl GeneratingWorldSource {
+    pub fn from_arc<G>(generator: Arc<G>) -> Self
+    where
+        G: WorldGenerator + 'static,
+    {
+        let (send_pos, recv_pos) = unbounded::<ChunkPosition>();
+        let (send_gen, recv_gen) = unbounded::<LoadedChunk>();
+        for _ in 0..8 {
+            // FIXME: do not hardcode '8' as a magic number
```

Comment:
I'd recommend starting just one thread here that polls the channel, then invokes `rayon::spawn` for each chunk that needs to be generated. Right now, we're blocking the Rayon thread poll by waiting on channels, which can cause a deadlock